### PR TITLE
Refactors chem application, (hopefully) fixes foam for good

### PR DESCRIPTION
**Pull request by SapphicOverload**
# Document the changes in your pull request

Refactors chem application to be bitwise and allow multiple application methods at once without having to apply the reagent twice, and makes foam use both touch and vapor application. Also prevents foam from applying more than once per second, and fixes reagent duplication when using the vapor application method.

Chem application refactor ported from tgstation/tgstation#53164

Closes #19010

# Changelog

:cl:  
tweak: refactors chem application to allow multiple application methods at once
tweak: foam actually works again, and uses both touch and vapor application and can only apply once per second
tweak: added permeability checks to a few more chems that should have them
bugfix: fixed vapor application duplicating reagents
/:cl:

### CODE REVIEW

**Comment by ToasterBiome** [comment]
Is this tested

**Comment by iamthesilly** [comment]
> Is this tested

Yes, works as advertised

File: code/game/objects/effects/effect_system/fluid_spread/effects_foam.dm [hunk 1]
```diff
@@ -117,11 +118,15 @@
...
+	if(foaming.foam_delay)
+		return
 
 	delta_time = min(delta_time SECONDS, lifetime)
 	var/fraction = (delta_time * MINIMUM_FOAM_DILUTION) / (initial(lifetime) * max(MINIMUM_FOAM_DILUTION, group.total_size))
-	reagents.reaction(foaming, TOUCH, fraction)
+	reagents.reaction(foaming, TOUCH|VAPOR, fraction)
 	lifetime -= delta_time
+	foaming.foam_delay = TRUE
+	addtimer(VARSET_CALLBACK(foaming, foam_delay, FALSE), 1 SECONDS)
```
Comment (ThatLing) [inline code/game/objects/effects/effect_system/fluid_spread/effects_foam.dm:129]: I feel like this should take ``delta_time`` into consideration

Comment (iamthesilly) [reply code/game/objects/effects/effect_system/fluid_spread/effects_foam.dm:129]: I'll see what I can do

Comment (iamthesilly) [reply code/game/objects/effects/effect_system/fluid_spread/effects_foam.dm:129]: I'm not sure I can find a solution that both takes delta_time into consideration, won't take disproportionally long if it happens on a longer or shorter tick than normal, and works across multiple sources of foam.

Comment (iamthesilly) [reply code/game/objects/effects/effect_system/fluid_spread/effects_foam.dm:129]: Actually from what I can tell I'm pretty sure timers take lag into account so it should be fine to keep it as is.

File: code/modules/reagents/chemistry/readme.md [hunk 2]
```diff
@@ -116,10 +116,10 @@ The holder (reagents datum) is the datum that holds a list of all reagents curre
 # About Reagents:
 Reagents are all the things you can mix and fille in bottles etc. This can be anything from rejuvs over water to ... iron. Each reagent also has a few procs - i'll explain those below.
 ```
-		reaction_mob(mob/living/L, method=TOUCH)
+		reaction_mob(var/mob/living/L, var/method=TOUCH)
```
Comment (JohnFulpWillard) [inline code/modules/reagents/chemistry/readme.md]: Undo this please lol (and the one above), we don't want to encourage uses of ``var/`` in args.

**Review by JohnFulpWillard — APPROVED** [review]

File: code/modules/reagents/chemistry/readme.md (hunk 2 above)
Comment (iamthesilly) [reply code/modules/reagents/chemistry/readme.md]: Alright, done

**Review by JohnFulpWillard — APPROVED** [review]

**Pull request merged**